File: C++/100_The3n+1problem.cpp
#include<cstdio>
#include<map>
using namespace std;
map<int,int>cycleLength;
int tempInt;
int calLength(int num){
    if(cycleLength.count(num)>0){
        return cycleLength[num];
    }
    else{
        if(num%2==1){
            tempInt=calLength(3*num+1)+1;
            cycleLength[num]=tempInt;
            return tempInt;
        }
        else{
            tempInt=calLength(num/2)+1;
            cycleLength[num]=tempInt;
            return tempInt;
        }
    }
}
int main(){
    int low,upp,maxx,i;
    cycleLength[1]=1;
    while(~scanf("%d%d",&low,&upp)){
       maxx=-1;
       if(low>upp){
            for(i=upp;i<=low;i++){
               tempInt=calLength(i);
               if(maxx<tempInt){
                    maxx=tempInt;
               }
            }
       }
       else{
            for(i=low;i<=upp;i++){
               tempInt=calLength(i);
               if(maxx<tempInt){
                    maxx=tempInt;
               }
            }
       }
       printf("%d %d %d\n",low,upp,maxx);
    }
    return 0;
}

File: Java/100_The 3n + 1 problem.java
import java.util.Map;
import java.util.HashMap;
import java.util.Scanner;

public class Main {
    public static Map<Integer, Integer> cycleLength = new HashMap<>();
    private static int calLength(int num){
        if(cycleLength.containsKey(num)){
            return cycleLength.get(num);
        }
        else{
            int tmpInt;
            if(num%2==1){
                tmpInt=calLength(3*num+1)+1;
                cycleLength.put(num,tmpInt);
                return tmpInt;
            }
            else{
                tmpInt=calLength(num/2)+1;
                cycleLength.put(num,tmpInt);
                return tmpInt;
            }
        }
    }
    public static void main(String[] args){
        cycleLength.put(1,1);
        Scanner input=new Scanner(System.in);
        while(input.hasNext()){
            int max=-1;
            int low=input.nextInt();
            int upp=input.nextInt();
            if(low>upp){
                for(int i=upp;i<=low;i++){
                    int tmpInt=calLength(i);
                    if(max<tmpInt){
                        max=tmpInt;
                    }
                }
            }
            else{
                for(int i=low;i<=upp;i++){
                    int tmpInt=calLength(i);
                    if(max<tmpInt){
                        max=tmpInt;
                    }
                }
            }

            System.out.format("%d %d %d%n",low,upp,max);
        }
    }
}
